File: codeforces/round49/MouseHunt.py
n = int(input())
c = list(map(int,input().split()))
a = list(map(int,input().split()))



def dfs(ADj,c,i,k,flag):
    flag[i][0] = 1
    v = ADj[i] - 1
    if flag[v][1] != 0:
        flag[i][1] = flag[v][1]
    else:
        flag[v][1] = k
    knum = flag[i][1]
    mincost = c[i]
    if flag[v][0] == 0:
        (mincost,knum) = dfs(ADj,c,v,k,flag)
    elif flag[v][0] == 1: # ring
        pass
    else:
        pass
    mincost = min(mincost,c[i])
    flag[i][1] = knum
    return (mincost,flag[i][1]) 

flag = [[0,0] for i in range(0,n)]
k = 1
result = [0 for i in range(0,n+1)]
for i in range(0,n):
    for j in range(0,n):
        flag[j][0] = 0
    #flag[i][1] = k
    (mincost,knum) = dfs(a,c,i,k,flag)
    #print(mincost,end=' ')
    #print(knum)
    result[knum] = max(result[knum],mincost)
    if knum == k:
        k += 1
        
#print(result)
print(sum(result))

File: codeforces/round49/MouseHunt.cpp
#include <bits/stdc++.h>

using namespace std;

#define N 200005
#define MIN(a,b) ((a<b)?a:b)
#define MAX(a,b) ((a<b)?b:a)

typedef struct{
    int num;
    void* next;
}Node;

int main(){
    int n =0;
    scanf("%d",&n);
    int c[N];
    int a[N];
    int r[N][3]; // ring array
    int result = 0;
    int path[N] = {0};
    for (int i = 1; i < n+1; i++) scanf("%d",c+i);
    for (int i = 1; i < n+1; i++) scanf("%d",a+i);
    for (int i=1; i<n+1;i++){
        int j = i;
        int mincost = c[j];
        path[0] = j;
        int path_len = 1;
        while(1){
            int ring = 0;
            //int ringmin = 2e4+5;
            for(int k = path_len-1; k >=0 ; --k){
                //ringmin = min(ringmin,path[k]);
                if(a[j] == path[k]){ // ring
                    ring = 1;
                    break;
                }  
            }
            if(ring) break;
            mincost = MIN(mincost,c[a[j]]);
            path[path_len++] = a[j];
            j = a[j];
        }
        printf("i = %d mincost = %d\n",i,mincost);
        result = MAX(result,mincost); 
    }
    printf("%d\n",result);
    return 0;
}
